### ISSUE TITLE
Database Recovery - Apcontact

### ISSUE THREAD

**Issue by shiribailem** [friendica/friendica#13358]
My backups failed and I had database corruption, the only loss being the majority of the apcontact table.

Looking at it, it looks like it's mostly just a mapping of all visible users and their profile data.

Will this content restore on it's own? Is there a way to trigger a restore if not? It looks like it should be entirely rebuildable from scratch, but I don't know if there are any measures that attempt to do so.

I'm hoping I don't have to scrap the whole database for this because everything else works fine and the server is online, just has issues looking up profiles from search.

**Comment by shiribailem** [comment]
Database backups are fixed going forward with multiple backup methods now.

As far as how I restored, I was able to almost all of the database. Everything was able to be read and dumped correctly *except* the apcontact table, of which only about the first 50k rows could be read.

This appeared to be a disk failure, so the cause isn't friendica related.

Posting here because restoring apcontact is a good general knowledge item, especially because in cases of corruption it's the most likely to become corrupt being the largest table by a wide margin (unless you're storing files in the database).

**Comment by MrPetovan** [comment]
There's no manual way to rebuild the `apcontact` table but on the bright side we don't use the `apcontact.id` field as a foreign key, so you didn't lose any irrecoverable data. I expect it to rebuild itself as the relevant contacts will be probed again over time.

**Comment by shiribailem** [comment]
Thanks for the confirmation, after this I at some point hit Check Database Table and I think it may have been something going on there because a lot of the issues disappeared immediately.

I figured even a "will resolve itself over time" would probably be good to just have in a searchable place for other admins.

For others doing database recovery, here's what I did to help others save time:

First I put the database in innodb recovery mode, and tried a regular dump. It of course kept failing and crashing on the apcontact table.

Then:
`mysqldump friendica --lock-tables --ignore-table=friendica.apcontact > friendica.sql`
Followed by doing a dump specifically of apcontact as much as possible `mysqldump friendica apcontact --lock-tables -f > friendica.apcontact.sql` (this gets at least some of it)

Then I wiped the database, rebuilt from there importing those two files. Once I had the user accounts set back up, everything opened up just fine.

I suppose this also means if someone wants to decrease the size of their backups they can probably get away with skipping that table.
